### PR TITLE
B/control namespace database api refactoring

### LINKED ISSUE #504: Continue backend refactoring for control database namespace
**Is your feature request related to a problem? Please describe:**  
We still need to finish refactoring our backend.

**Please describe the solution you'd like:**  
The solution is already implemented for the `workload` API namespace in #446 .

**Please describe alternatives you've considered:**  
None.

**Affected Component(s):**  
`CockpitBackend`

### PR DESCRIPTION
**Pull request by Alexander-Dubrawski**
# Resolves #504 

**Does your pull request add a feature? Please describe:**  
This PR refactors the `control` API namespace. It creates a new `database` namespace. Moreover, we have now a simple `socket_manager`. From it, we can import a `ManagerSocket`, or `GeneratorSocket`. We can use them witch the context manager. It takes care to open and close a socket connection.

**Affected Component(s):**  
`App`

### CODE REVIEW

**Comment by heseding** [comment]
How about we break the namespace into several namespaces, e.g. `/database`, `/data`, etc.?

**Comment by Alexander-Dubrawski** [comment]
> How about we break the namespace into several namespaces, e.g. `/database`, `/data`, etc.?

I was thinking about the same. But I think we shouldn't do it for every endpoint. We need to find some  *logic modules*.  One example of such a module is the workload namespace. Maybe: 

* Plugin 
* Database 
* SQL 

@fabianhe what do you think?

**Comment by heseding** [comment]
> I think we shouldn't do it for every endpoint. We need to find some _logic modules_.

Definitely, we could put it on the agenda for our next backend meeting?

>  One example of such a module is the workload namespace. Maybe:
> 
> * Plugin
> * Database
> * SQL
> 
> @fabianhe what do you think?

Sounds great to me!

File: hyrisecockpit/api/app/database/controller.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    """Controller for access and register databases."""
+
+    @responds(
+        schema=DetailedDatabaseSchema(
+            many=True, only=("id", "host", "port", "number_workers", "dbname",)
+        ),
+        api=api,
+    )
+    def get(self) -> List[DetailedDatabase]:
+        """Get all Workloads."""
```
Comment (heseding) [inline hyrisecockpit/api/app/database/controller.py]:
```suggestion
        """Get all databases."""
```

File: hyrisecockpit/api/app/database/controller.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return DatabaseService.get_databases()
+
+    @accepts(schema=DetailedDatabaseSchema, api=api)
+    def post(self) -> Response:
+        """Register new database."""
+        status_code = DatabaseService.register_database(request.parsed_obj)
```
Comment (heseding) [inline hyrisecockpit/api/app/database/controller.py]: Just for typing:
```suggestion
        interface: DetailedDatabaseSchema = request.parsed_obj
        status_code = DatabaseService.register_database(interface)
```

File: hyrisecockpit/api/app/database/controller.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        return Response(status=status_code)
+
+    @accepts(schema=DatabaseSchmea, api=api)
```
Comment (heseding) [inline hyrisecockpit/api/app/database/controller.py]:
```suggestion
    @accepts(schema=DatabaseSchema, api=api)
```

File: hyrisecockpit/api/app/database/interface.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+class DatabaseInterface(TypedDict):
+    """Interface of a Database."""
+
+    id: str
+
+
+class DetailedDatabaseInterface(TypedDict):
+    """Interface of a detailed database."""
+
+    id: str
```
Comment (heseding) [inline hyrisecockpit/api/app/database/interface.py]:
```suggestion
class DetailedDatabaseInterface(DatabaseInterface):
    """Interface of a detailed database."""

```

File: hyrisecockpit/api/app/routes.py [hunk 5]
```diff
@@ -7,7 +7,7 @@
 def register_routes(api: Api, app: Flask, root: str = "/api") -> None:
     """Register all sub-routes."""
     from .workload import register_routes as attach_workload
-    from .control import register_routes as attach_control
+    from .database import register_routes as attach_control
```
Comment (heseding) [inline hyrisecockpit/api/app/routes.py]:
```suggestion
    from .database import register_routes as attach_database
```

File: hyrisecockpit/api/app/socket_manager.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    def send_req(self, message: Request) -> Response:
+        """Send message to socket."""
+        self._socket.send_json(message)
+        response: Response = self._socket.recv_json()
+        return response
+
+
+class GeneratorSocket:
+    """GeneratorSocket that sands request to the generator."""
```
Comment (heseding) [inline hyrisecockpit/api/app/socket_manager.py]:
```suggestion
    """GeneratorSocket that sends requests to the generator."""
```

File: hyrisecockpit/api/app/socket_manager.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        self._socket.close()
+        return None
+
+    def send_message(self, message: Request) -> Response:
+        """Send message to generator."""
+        return self._socket.send_req(message)
+
+
+class ManagerSocket:
+    """ManagerSocket that sands request to the Manager."""
```
Comment (heseding) [inline hyrisecockpit/api/app/socket_manager.py]:
```suggestion
    """ManagerSocket that sends requests to the Manager."""
```

File: hyrisecockpit/database_manager/driver.py [hunk 8]
```diff
@@ -52,3 +52,11 @@ def _create_connection_pool(self, n_connections: int) -> pool:
...
         """Return the connection pool."""
         return self._connection_pool
+
+    def get_password(self) -> str:
+        """Return database password."""
+        return self._password
+
+    def get_user(self) -> str:
+        """Return database user."""
+        return self._user
```
Comment (heseding) [inline hyrisecockpit/database_manager/driver.py]: We could either make `._password` and `._user` explicitly public (e.g. `.password`) and use it directly, or declare this as a getter and name it `password` (see https://docs.python.org/3/howto/descriptor.html).

File: hyrisecockpit/api/app/database/schema.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+from hyrisecockpit.api.app.database.model import (
+    AvailableBenchmarkTables,
+    BenchmarkTables,
+    Database,
+    DetailedDatabase,
+)
+
+
+class DatabaseSchmea(Schema):
```
Comment (PeterTsayun) [inline hyrisecockpit/api/app/database/schema.py]:
```suggestion
class DatabaseSchema(Schema):
```

File: tests/api/database/test_schema.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        self, detailed_database_schema: DetailedDatabaseSchema
+    ) -> None:
+        """A detailed database schema can be created."""
+        assert detailed_database_schema
+
+    def test_creates_available_benchmark_tables_schema(
+        self, available_benchmark_tables_schema: AvailableBenchmarkTablesSchema
+    ) -> None:
+        """A available benchmark tables schema can be created."""
+        assert AvailableBenchmarkTablesSchema
```
Comment (PeterTsayun) [inline tests/api/database/test_schema.py]:
```suggestion
        assert available_benchmark_tables_schema
```

File: hyrisecockpit/api/app/database/controller.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+class Worker(Resource):
+    """Manage start and stop of worker pool at all databases."""
+
+    def post(self) -> Response:
+        """Start worker pool for all databases."""
+        status_code = DatabaseService.start_worker_pool()
+        return Response(status=status_code)
+
+    def delete(self) -> Response:
+        """Start worker pool for all databases."""
```
Comment (PeterTsayun) [inline hyrisecockpit/api/app/database/controller.py:89]: What is the use case of starting/stopping the workers separately from the workload execution?

File: hyrisecockpit/database_manager/manager.py [hunk 12]
```diff
@@ -136,6 +136,8 @@ def _call_get_databases(self, body: Body) -> Response:
                 "port": database.driver.port,
                 "number_workers": database.number_workers,
                 "dbname": database.driver.dbname,
+                "user": database.driver.user(),
+                "password": database.driver.password(),
```
Comment (PeterTsayun) [inline hyrisecockpit/database_manager/manager.py]: Do we actually need authentification information for the Hyrise instances?

Comment (Alexander-Dubrawski) [reply hyrisecockpit/database_manager/manager.py]: It is cleaner to give the hole information to the API. With schemas, we can specify which information to give to the outside. But all in all we actually just need the host and port information to connect to the hyrise 🌝 .

File: hyrisecockpit/api/app/database/controller.py (hunk 11 above)
Comment (Alexander-Dubrawski) [reply hyrisecockpit/api/app/database/controller.py:89]: It is a cleaner logical. In the future, we could for example just add and remove workers on the fly.

**Review by PeterTsayun — APPROVED** [review]

**Pull request merged**